src/backend/access/transam/varsup.c

```diff
@@ -584,20 +584,40 @@ GetNewObjectId(void)
 }
 
 /*
- * AdvanceObjectId -- advance object id counter for QE nodes
+ * AdvanceObjectId -- advance object id counter for QD and QE nodes
  *
- * The QD provides the preassigned OID to the QE nodes which will be
- * used as the relation's OID. QE nodes do not use this OID as the
- * relfilenode value anymore so the OID counter is not
- * incremented. This function forcefully increments the QE node's OID
- * counter to be about the same as the OID provided by the QD node.
+ * When advancing the Oid counter of a QD, it should only be for the purpose
+ * of syncing Oid counters logically compared with the numeric maximum Oid
+ * counter value among the primary segments.
+ *
+ * When advancing the Oid counter of a QE, the QD provides the preassigned OID
+ * to the QE nodes which will be used as the relation's OID. QE nodes do not
+ * use this OID as the relfilenode value anymore so the OID counter is not
+ * incremented. This function forcefully increments the QE node's OID counter
+ * to be about the same as the OID provided by the QD node.
  */
 void
 AdvanceObjectId(Oid newOid)
 {
 	LWLockAcquire(OidGenLock, LW_EXCLUSIVE);
 
-	while(GetNewObjectIdUnderLock() <= newOid);
+	if (OidFollowsNextOid(newOid))
+	{
+		/* Determine if we need to write an xlog */
+		bool writexlog = (int32)(newOid - ShmemVariableCache->nextOid) >= ShmemVariableCache->oidCount;
+
+		/*
+		 * We directly set the nextOid counter to the preassigned OID instead
+		 * of doing increments. Update the oidCount and create an XLog (only
+		 * if we have exhausted the current oidCount) as if we were doing
+		 * GetNewObjectIdUnderLock(). We should always be moving forward and
+		 * never backwards.
+		 */
+		ShmemVariableCache->nextOid = newOid;
+		ShmemVariableCache->oidCount = VAR_OID_PREFETCH - (ShmemVariableCache->nextOid % VAR_OID_PREFETCH);
```

Comment:
No it doesn't really need to be a multiple of `VAR_OID_PREFETCH`.

```suggestion
		ShmemVariableCache->oidCount = VAR_OID_PREFETCH;
```

A `VAR_OID_PREFETCH` would work just as fine. On a second thought, I finally see what you're suggesting! @hlinnaka

Reply:
Yeah, makes sense.  I was being very conservative on setting the `oidCount` to somewhat match the logic as if we were doing incremental calls to `getNewObjectId()`.  Updated the code to make this less convoluted.